File: packages/example_app/lib/features/product/data/repositories/product_repository_fake.dart
import 'package:dartz/dartz.dart';

import '../../../../core/errors/failures.dart';
import '../../../../core/pagination/pagination.dart';
import '../../domain/entities/product_entity.dart';
import '../../domain/repositories/product_repository.dart';
import '../models/product_model.dart';

/// In-memory Product repository for development, demos, and tests.
///
/// State is held in a private list — call [seeded] to bootstrap with mock
/// data, or pass [initial] explicitly.
class ProductRepositoryFake implements ProductRepository {
  ProductRepositoryFake({List<ProductModel>? initial})
      : _items = [...?initial];

  factory ProductRepositoryFake.seeded({int count = 10}) =>
      ProductRepositoryFake(initial: ProductModel.dummyList(count));

  final List<ProductModel> _items;
  var _nextId = 1000;

  @override
  Future<Either<Failure, List<ProductEntity>>> getAll() async {
    return Right(_items.map((m) => m.toEntity()).toList(growable: false));
  }

  @override
  Future<Either<Failure, PaginatedResponse<ProductEntity>>> getAllPaginated(
    PaginationParams params,
  ) async {
    final start = params.offset.clamp(0, _items.length);
    final end = (params.offset + params.limit).clamp(0, _items.length);
    final slice = _items
        .sublist(start, end)
        .map((m) => m.toEntity())
        .toList(growable: false);
    return Right(PaginatedResponse<ProductEntity>(
      items: slice,
      total: _items.length,
      offset: params.offset,
      limit: params.limit,
    ));
  }

  @override
  Future<Either<Failure, ProductEntity>> getById(String id) async {
    final i = _items.indexWhere((m) => m.id == id);
    if (i < 0) return Left(NotFoundFailure('No Product with id $id'));
    return Right(_items[i].toEntity());
  }

  @override
  Future<Either<Failure, ProductEntity>> add(ProductEntity entity) async {
    final assignedId =
        entity.id.isEmpty ? 'product_${_nextId++}' : entity.id;
    final model = ProductModel.fromEntity(entity.copyWith(id: assignedId));
    _items.add(model);
    return Right(model.toEntity());
  }

  @override
  Future<Either<Failure, ProductEntity>> update(ProductEntity entity) async {
    final i = _items.indexWhere((m) => m.id == entity.id);
    if (i < 0) {
      return Left(NotFoundFailure('No Product with id ${entity.id}'));
    }
    final model = ProductModel.fromEntity(entity);
    _items[i] = model;
    return Right(model.toEntity());
  }

  @override
  Future<Either<Failure, Unit>> delete(String id) async {
    final before = _items.length;
    _items.removeWhere((m) => m.id == id);
    if (_items.length == before) {
      return Left(NotFoundFailure('No Product with id $id'));
    }
    return const Right(unit);
  }

  @override
  Future<Either<Failure, List<ProductEntity>>> search(String query) async {
    final q = query.toLowerCase();
    final hits = _items
        .where((m) =>
            (m.name ?? '').toLowerCase().contains(q) ||
            (m.description ?? '').toLowerCase().contains(q))
        .map((m) => m.toEntity())
        .toList(growable: false);
    return Right(hits);
  }
}

File: packages/example_react_app/src/features/product/data/repositories/product.repository-fake.ts
/// In-memory fake repository for Product.
import type { Either } from '../../../../core/usecase/usecase';
import { left, right } from '../../../../core/usecase/usecase';
import { NotFoundFailure, type Failure } from '../../../../core/errors/failures';
import type { ProductEntity } from '../../domain/entities/product.entity';
import type { ProductRepository } from '../../domain/repositories/product.repository';
import { toEntity, dummyProductList } from '../models/product.model';
import type { PaginationParams, PaginatedResponse } from '../../../../core/pagination/pagination';

export class ProductRepositoryFake implements ProductRepository {
  private data: ProductEntity[] = [];

  constructor(seedData: ProductEntity[] = []) {
    this.data = seedData;
  }

  static seeded(count: number = 5): ProductRepositoryFake {
    const models = dummyProductList(count);
    return new ProductRepositoryFake(models.map(toEntity));
  }

  async getAll(): Promise<Either<Failure, ProductEntity[]>> {
    return right([...this.data]);
  }

  async getById(id: string): Promise<Either<Failure, ProductEntity>> {
    const entity = this.data.find((e) => e.id === id);
    if (!entity) {
      return left(new NotFoundFailure(`Product with id ${id} not found`));
    }
    return right(entity);
  }

  async add(entity: ProductEntity): Promise<Either<Failure, ProductEntity>> {
    const newEntity = { ...entity, createdAt: new Date(), updatedAt: new Date() };
    this.data.push(newEntity);
    return right(newEntity);
  }

  async update(entity: ProductEntity): Promise<Either<Failure, ProductEntity>> {
    const index = this.data.findIndex((e) => e.id === entity.id);
    if (index === -1) {
      return left(new NotFoundFailure(`Product with id ${entity.id} not found`));
    }
    const updated = { ...entity, updatedAt: new Date() };
    this.data[index] = updated;
    return right(updated);
  }

  async delete(id: string): Promise<Either<Failure, void>> {
    const index = this.data.findIndex((e) => e.id === id);
    if (index === -1) {
      return left(new NotFoundFailure(`Product with id ${id} not found`));
    }
    this.data.splice(index, 1);
    return right(undefined);
  }

  async search(query: string): Promise<Either<Failure, ProductEntity[]>> {
    const lowerQuery = query.toLowerCase();
    const results = this.data.filter(
      (e) =>
        e.name?.toLowerCase().includes(lowerQuery) ||
        e.description?.toLowerCase().includes(lowerQuery)
    );
    return right(results);
  }

  async getAllPaginated(params: PaginationParams): Promise<Either<Failure, PaginatedResponse<ProductEntity>>> {
    const { offset, limit } = params;
    const items = this.data.slice(offset, offset + limit);
    return right({
      items,
      total: this.data.length,
      offset,
      limit,
    });
  }
}
